EKF/control.cpp

```diff
@@ -895,21 +895,34 @@ void Ekf::checkVerticalAccelerationHealth()
 	// Clipping causes the average accel reading to move towards zero which makes the INS
 	// think it is falling and produces positive vertical innovations
 
-	const float var_product_lim = sq(_params.vert_innov_test_lim) * sq(_params.vert_innov_test_lim);
-	const bool is_fusing_gps_vel = !_gps_hgt_intermittent;
-	const bool is_fusing_baro_alt = _control_status.flags.baro_hgt && !_baro_hgt_faulty;
-	const bool are_vertical_pos_and_vel_independant = is_fusing_gps_vel && is_fusing_baro_alt; // TODO: should we add range hgt here?
 	const float pos_vel_innov_product = _gps_pos_innov(2) * fmaxf(fabsf(_gps_vel_innov(2)),fabsf(_ev_vel_innov(2)));
 	const float pos_vel_innov_var_product = _gps_pos_innov_var(2) * fmaxf(fabsf(_gps_vel_innov_var(2)),fabsf(_ev_vel_innov_var(2)));
+	const float var_product_lim = sq(_params.vert_innov_test_lim) * sq(_params.vert_innov_test_lim);
 	const bool are_pos_vel_sensor_in_agreement = sq(pos_vel_innov_product) > var_product_lim * (pos_vel_innov_var_product);
 
 	// A positive innovation indicates that the inertial nav thinks it is falling
 	// This is caused by average of the Z accelerometer being 0, due to clipping
 	const bool is_inertial_nav_falling = _gps_vel_innov(2) > 0.0f || _ev_vel_innov(2) > 0.0f;
 
-	const bool bad_vert_accel = are_vertical_pos_and_vel_independant &&
-			      are_pos_vel_sensor_in_agreement &&
-			      is_inertial_nav_falling;
+	// If vertical position and velocity come from independent sensors then we can
+	// trust them more if they disagree with the IMU
+	const bool are_vertical_pos_and_vel_independant = !(_control_status.flags.gps_hgt && _control_status.flags.gps) &&
+							  !(_control_status.flags.ev_hgt && _control_status.flags.ev_vel);
```

Comment:
Unfortunately, if there is only a baro or if GPS aiding is active but we detected bad data, this condition will be true. This is why we had the tests `is_fusing_baro_alt` and `is_fusing_gps_vel`

Reply:
It is getting messy and hard to follow since adding support for different combinations of velocity and height, so it needs to be abstracted a different way. Will rework.